routing/index_graph_serializer.cpp

```diff
@@ -0,0 +1,82 @@
+#include "routing/index_graph_serializer.hpp"
+
+namespace routing
+{
+// static
+uint8_t constexpr IndexGraphSerializer::kVersion;
+
+void IndexGraphSerializer::SectionSerializer::SerializeToBuffer(
+    IndexGraph const & graph, unordered_map<uint32_t, VehicleMask> const & masks,
+    JointIdEncoder & jointEncoder)
+{
+  m_buffer.clear();
+  MemWriter<vector<uint8_t>> memWriter(m_buffer);
+  BitWriter<MemWriter<vector<uint8_t>>> writer(memWriter);
+
+  // -1 for uint32_t is some confusing, but it allows process first iteration in the common way.
+  // It works because uint32_t is modular ring type.
+  uint32_t prevFeatureId = -1;
+
+  for (uint32_t const featureId : m_featureIds)
+  {
+    RoadJointIds const & road = graph.GetRoad(featureId);
+
+    uint32_t const featureDelta = featureId - prevFeatureId;
+    Write32(writer, featureDelta);
+    prevFeatureId = featureId;
+
+    uint32_t const pointCap = road.GetMaxPointId() + 1;
+    Write32(writer, pointCap);
+
+    uint32_t prevPointId = -1;
+    road.ForEachJoint([&](uint32_t pointId, Joint::Id jointId) {
+      uint32_t const pointDelta = pointId - prevPointId;
+      Write32(writer, pointDelta);
+      prevPointId = pointId;
+
+      jointEncoder.Write(jointId, writer);
+    });
+  }
+}
+
+// static
+VehicleMask IndexGraphSerializer::GetRoadMask(unordered_map<uint32_t, VehicleMask> const & masks,
+                                              uint32_t featureId)
+{
+  auto const & it = masks.find(featureId);
+  if (it == masks.cend())
```

Comment:
On the generator stage all such things must be CHECKED, as there's no graceful error fixing on the stage.